crypto/bio/b_sock.c

```diff
@@ -83,6 +83,26 @@ NETDB_DEFINE_CONTEXT
 static int wsa_init_done = 0;
 # endif
 
+# if defined(__GLIBC__)
+#  if defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 19)
+#    if defined(_DEFAULT_SOURCE)
+#     define HAVE_GETHOSTBYNAME_R
+#    endif
+#   else
+#    if defined(_BSD_SOURCE) || defined(_SVID_SOURCE)
+#     define HAVE_GETHOSTBYNAME_R
+#    endif
+#   endif
+#  else
+#   if defined(_BSD_SOURCE) || defined(_SVID_SOURCE)
+#    define HAVE_GETHOSTBYNAME_R
+#   endif
+#  endif
+# endif
```

Comment:
You just define _DEFAULT_SOURCE and _BSD_SOURCE instead, and then assume that it's available.

Reply:
Checking the a define like _DEFAULT_SOURCE does not make sense, it's something you need to set.

Reply:
Please see https://github.com/openssl/openssl/issues/7228#issuecomment-421881712

Reply:
Uhmmm @kroeckx, do you assume that `_DEFAULT_SOURCE` or `_BSD_SOURCE` are valid everywhere?

Reply:
From my man file for feature_test_macros:

       If no feature test macros are explicitly defined,  then  the  following
       feature  test macros are defined by default: _BSD_SOURCE, _SVID_SOURCE,
       _POSIX_SOURCE, and _POSIX_C_SOURCE=200909L (200112L in  glibc  versions
       before  2.10;  199506L  in  glibc versions before 2.4; 199309L in glibc
       versions before 2.1).

Sounds like we don't need to define these as long as GLIBC of the requisite versions exist.

Reply:
@paulidale appears to be correct as far as I can tell.

Reply:
I think there is still a misunderstanding on what this means. glibc always had this function. But to make sure that the function is visible, in earlier versions you had to define either _BSD_SOURCE or _SVID_SOURCE, and in later versions _DEFAULT_SOURCE.

Those defines can also be defined by default. In current versions, _DEFAULT_SOURE is defined by default, but is undefined if you define any of __STRICT_ANSI__,  _ISOC99_SOURCE, _POSIX_SOURCE, _POSIX_C_SOURCE, _XOPEN_SOURCE, _XOPEN_SOURCE_EXTENDED  or _SVID_SOURCE. The compiler can set some of those options depending on the flags you pass.

Anyway, what it comes down to is that unless someone passes an options to the compiler, you get it by default on glibc, and there is no need to do all that checking. If they pass some option that turns off _DEFAULT_SOURCE, defining _DEFAULT_SOURE in the file before the includes will give it to you.

So adding a define of _DEFAULT_SOURE and _BSD_SOURCE before the includes will guarantee that the function is available.

Reply:
Why should we be overriding the settings of the people compiling the code? If they turn off _DEFAULT_SOURCE which would cause gethostbyname_r to not exist then why should we force it back on again?

Reply:
Because it's not the intention to disable gethostbyname_r(), while the code has the intention to use it when it's available.

I really don't understand why it's so hard to just add 2 defines instead of this complicated logic.

Reply:
I mean on glibc, not any other libc.

Reply:
But it's also available on things AIX and HP/UX, but with different parameters.

Reply:
We should fallback to non-use rather than fallback to use. The existing macro covers testing for glibc and it shouldn't be the responsibility of this pull request to enable usage of gethosbyname_r on all possible platforms. At least that's my opinion.

Reply:
This is not a bug fix, this is a performance improvement. It does not need to be "fixed" on all OSes.